flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/resource/UserResourceManager.java

```diff
@@ -0,0 +1,175 @@
+                            + "in the current version, skipping.",
+                    resourceUri.getUri());
+            return;
+        }
+
+        Path path = new Path(resourceUri.getUri());
+        // check resource firstly
+        checkResource(resourceUri.getResourceType(), path);
+
+        URL localUrl;
+        // check resource scheme
+        String scheme = StringUtils.lowerCase(path.toUri().getScheme());
+        // download resource to local path firstly if in remote
+        if (scheme != null && !"file".equals(scheme)) {
+            localUrl = downloadResource(path);
+        } else {
+            localUrl = getURLFromPath(path);
+        }
+
+        // only need add jar resource to classloader
+        if (ResourceType.JAR.equals(resourceUri.getResourceType())) {
+            userClassLoader.addURL(localUrl);
+            LOG.info("Added jar resource [{}] to class path.", localUrl.toString());
+        }
+
+        resourceInfos.put(resourceUri, localUrl);
+        LOG.info("Register resource [{}] successfully.", resourceUri.getUri());
+    }
+
+    public Map<ResourceUri, URL> getResources() {
+        return Collections.unmodifiableMap(resourceInfos);
+    }
+
+    public Set<URL> getJarResourceURLs() {
+        return resourceInfos.entrySet().stream()
+                .filter(entry -> ResourceType.JAR.equals(entry.getKey().getResourceType()))
+                .map(Map.Entry::getValue)
+                .collect(Collectors.toSet());
+    }
+
+    private void checkResource(ResourceType resourceType, Path path) {
+        try {
+            FileSystem fs = path.getFileSystem();
+            // check resource exists firstly
+            if (!fs.exists(path)) {
+                throw new IllegalArgumentException(String.format("Resource [%s] not found.", path));
+            }
+
+            // register directory is not allowed for jar resource
+            if (ResourceType.JAR.equals(resourceType) && fs.getFileStatus(path).isDir()) {
+                throw new IllegalArgumentException(
+                        String.format(
+                                "Directory [%s] is not allowed for registering jar resource.",
+                                path));
+            }
+
+        } catch (IOException e) {
+            throw new FlinkRuntimeException(
+                    String.format("Failed to check resource [%s].", path), e);
+        }
+    }
+
+    @VisibleForTesting
+    protected URL downloadResource(Path remotePath) {
+        final String localTmpDir = config.get(TableConfigOptions.DOWNLOADED_RESOURCES_TMP_DIR);
+        // add an uuid prefix for local file name
+        Path localPath =
+                new Path(
+                        String.format(
+                                "%s%s%s-%s",
+                                localTmpDir,
+                                Path.SEPARATOR,
+                                UUID.randomUUID(),
+                                remotePath.getName()));
```

Comment:
After some thought, I think we should not add UUID suffix or prefix on resource name, we should add an sudDir which created with UUID suffix. For example, In gateway, every session need to have it owner unique subDir, this can be controlled by `RESOURCE_DOWNLOAD_DIR` option. Hive is done so.

Reply:
Even though every session has a unique dir, there may be still files with conflict file names but are different resources.